build.rs

```diff
@@ -1,34 +1,156 @@
+    }
+
+    // Allow specifying custom toolchain specifically for SDL2.
+    if let Ok(toolchain) = env::var("HWLOC_TOOLCHAIN") {
+        cfg.define("CMAKE_TOOLCHAIN_FILE", &toolchain);
+    } else {
+        // Override __FLTUSED__ to keep the _fltused symbol from getting defined in the static build.
+        // This conflicts and fails to link properly when building statically on Windows, likely due to
+        // COMDAT conflicts/breakage happening somewhere.
+        #[cfg(feature = "static-link")]
+        cfg.cflag("-D__FLTUSED__");
+
+        #[cfg(target_os = "linux")]
+        {
+            // Add common flag for affected version and above
+            use version_compare::{compare_to, Cmp};
+            if let Ok(version) = std::process::Command::new("cc")
+                .arg("-dumpversion")
+                .output()
+            {
+                let affected =
+                    compare_to(std::str::from_utf8(&version.stdout).unwrap(), "10", Cmp::Ge)
+                        .unwrap_or(true);
+                if affected {
+                    cfg.cflag("-fcommon");
+                }
+            }
+        }
+    }
+    cfg.build()
+}
+
 fn main() {
-    let required_version = if cfg!(feature = "hwloc-2_8_0") {
-        "2.8.0"
+    let git_dir = env::var("OUT_DIR").expect("No output directory given");
+    let (required_version, source_version) = if cfg!(feature = "hwloc-2_8_0") {
+        ("2.8.0", "v2.8")
     } else if cfg!(feature = "hwloc-2_5_0") {
-        "2.5.0"
+        ("2.5.0", "v2.5")
     } else if cfg!(feature = "hwloc-2_4_0") {
-        "2.4.0"
+        ("2.4.0", "v2.4")
     } else if cfg!(feature = "hwloc-2_3_0") {
-        "2.3.0"
+        ("2.3.0", "v2.3")
     } else if cfg!(feature = "hwloc-2_2_0") {
-        "2.2.0"
+        ("2.2.0", "v2.2")
     } else if cfg!(feature = "hwloc-2_1_0") {
-        "2.1.0"
+        ("2.1.0", "v2.1")
     } else if cfg!(feature = "hwloc-2_0_4") {
-        "2.0.4"
+        ("2.0.4", "v2.x")
     } else {
-        "2.0.0"
+        ("2.0.0", "v2.0")
     };
-    let lib = pkg_config::Config::new()
-        .atleast_version(required_version)
-        .statik(true)
-        .probe("hwloc")
-        .expect("Could not find a suitable version of hwloc");
-    if cfg!(target_family = "unix") {
-        for link_path in lib.link_paths {
-            println!(
-                "cargo:rustc-link-arg=-Wl,-rpath,{}",
-                link_path
-                    .to_str()
-                    .expect("Link path is not an UTF-8 string")
-            );
+    #[cfg(feature = "bundled")]
+    {
+        let mut hwloc2_source_path = PathBuf::from(git_dir);
+        fetch_hwloc(hwloc2_source_path.as_path(), source_version);
+        hwloc2_source_path.push("hwloc");
+        #[cfg(target_os = "windows")]
+        let cmake = hwloc2_source_path.join("contrib").join("windows-cmake").join("CMakeLists.txt").exists();
+        #[cfg(not(target_os = "windows"))]
+        let cmake = false;
```

Comment:
Here you are not using anything Windows-specific, so you can do...

```rust
let cmake = cfg!(target_os = "windows") && hwloc2_source_path.join("contrib").join("windows-cmake").join("CMakeLists.txt").exists();
```

...and this way, the code gets type-checked even when it is built on Linux (which is what the CI will usually do).

Reply:
On an unrelated note, do we really need to support CMake-less windows builds, given that hwloc has supported CMake since v2.8 and we can totally decide to use such a version in bundled builds?

Reply:
Probably not, my testing shows that the autotools based build only works on the -gnu toolchain, and for windows if the winapi crate is also used by the user then that won't work. The winapi crate only builds on the -msvc toolchain.

Reply:
I would just delete it then.